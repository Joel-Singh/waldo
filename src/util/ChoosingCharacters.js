import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { characterInformation } from "./constants";
import getFirebaseFunctions from "./firebase";

async function chooseAllCharactersIn(map) {
  await chooseMultipleCharacters(characterInformation[map].map(({displayName}) => (displayName)));
}

async function chooseMultipleCharacters(displayNameArr) {
  for (const displayName of displayNameArr) {
    await chooseCharacter(displayName);
  }
}

async function chooseCharacter(displayName) {
  if (!(await areFakeCoordsInDatabase()))
    throw new Error("No fake coords in database!")

  await chooseCharacterAtPosition(displayName, 0, 0);

  async function areFakeCoordsInDatabase() {
    const { getCharCoordsInDb } = getFirebaseFunctions()
    const charCoords = await getCharCoordsInDb()

    const coordsAreZero = ({x, y}) => x === 0 && y === 0
    const allCharCoordsAreZero = !Object.entries(charCoords).every(([, coords]) => coordsAreZero(coords))
    if (allCharCoordsAreZero)
      return false

    return true
  }
}

async function chooseCharacterAtPosition(displayName, xPos, yPos) {
  const gamescreen = screen.getByTestId("Gamescreen");
  userEvent.click(gamescreen, { screenX: xPos, screenY: yPos });

  const charBtn = screen.getByText(displayName);
  await act(async () => {
    userEvent.click(charBtn);

    const flushPromises = () =>
      new Promise((resolve) => setTimeout(resolve, 200));

    await flushPromises()
  });
}

export { chooseAllCharactersIn, chooseCharacter, chooseCharacterAtPosition }
