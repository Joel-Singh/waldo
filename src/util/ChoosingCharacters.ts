import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { act } from "react-dom/test-utils";
import { characterInformation } from "./constants";
import { MapName } from "./constants";

async function chooseAllCharactersIn(map: MapName) {
  await chooseMultipleCharacters(
    characterInformation[map].map(({ displayName }) => displayName),
    map,
  );
}

async function chooseMultipleCharacters(displayNameArr, map: MapName) {
  for (const displayName of displayNameArr) {
    await chooseCharacter(displayName, map);
  }
}

async function chooseCharacter(displayName, map: MapName) {
  const characterToChoose = characterInformation[map].find(
    (charInfo) => charInfo.displayName === displayName,
  );

  const { x, y } = characterToChoose.coords;
  await chooseCharacterAtPosition(displayName, x, y);
}

async function chooseCharacterAtPosition(displayName, xPos, yPos) {
  const gamescreen = screen.getByTestId("Gamescreen");
  userEvent.click(gamescreen, { screenX: xPos, screenY: yPos });

  const charBtn = screen.getByText(displayName);
  await act(async () => {
    userEvent.click(charBtn);

    const flushPromises = () =>
      new Promise((resolve) => setTimeout(resolve, 200));

    await flushPromises();
  });
}

export { chooseAllCharactersIn, chooseCharacter, chooseCharacterAtPosition };
