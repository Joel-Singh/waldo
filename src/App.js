import { Route, Routes } from "react-router-dom";
import SelectionScreen from "./components/SelectionScreen";
import { getGamescreens, getMapPreviews } from "./util/componentInstantiations";
import HighscoreScreen from "./components/HighscoreScreen"

// eslint-disable-next-line
import all from "./styles/all.css";
import { useState } from "react";

function App() {
  const [showHighscoreScreen, setShowHighscoreScreen] = useState(false);

  const {
    maze: mazeGamescreen,
    beach: beachGamescreen,
    snow: snowGamescreen,
  } = getGamescreens(() => setShowHighscoreScreen(true));

  const { mazeMapPreview, beachMapPreview, snowMapPreview } = getMapPreviews();
  return (
    <div className="App">
      <Routes>
        <Route path="*" element={null} />
        <Route
          path="/"
          element={
            <SelectionScreen
              mapPreviews={[mazeMapPreview, beachMapPreview, snowMapPreview]}
            />
          }
        />
        <Route path="/maze" element={!showHighscoreScreen ? mazeGamescreen : (<HighscoreScreen map="maze" />)} />
        <Route path="/beach" element={beachGamescreen} />
        <Route path="/snow" element={snowGamescreen} />
      </Routes>
    </div>
  );
}

export default App;
