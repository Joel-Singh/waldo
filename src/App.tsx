import { Route, Routes } from "react-router-dom";
import SelectionScreen from "./components/SelectionScreen";
import { getGamescreens, getMapPreviews } from "./util/componentInstantiations";
import FinalScoreScreen from "./components/FinalScoreScreen";

import "./styles/all.css";
import { useState } from "react";
import { allMaps } from "./util/constants";

function App() {
  const [showFinalScoreScreen, setShowFinalScoreScreen] = useState(false);
  const [currentPlayerScore, setCurrentPlayerScore] = useState(null);

  return (
    <div className="App">
      <Routes>
        <Route path="*" element={<div>Not a valid path</ div>} />
        <Route
          path="/"
          element={
            <SelectionScreen
              mapPreviews={allMaps.map((map) => getMapPreviews()[map])}
            />
          }
        />
        {allMaps.map((map) => createRouteForMap(map))}
      </Routes>
    </div>
  );

  function createRouteForMap(map) {
    const gamescreen = getGamescreens((timeElapsed) => {
      setShowFinalScoreScreen(true);
      setCurrentPlayerScore(timeElapsed);
    })[map];

    return (
      <Route
        path={`/${map}`}
        key={map}
        element={
          !showFinalScoreScreen ? (
            gamescreen
          ) : (
            <FinalScoreScreen
              finalScore={currentPlayerScore}
              backToSelectionScreenBtnOnClick={() =>
                setShowFinalScoreScreen(false)
              }
            />
          )
        }
      />
    );
  }
}

export default App;
