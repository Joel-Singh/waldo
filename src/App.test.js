import { render, screen } from "@testing-library/react";
import { MemoryRouter, BrowserRouter } from "react-router-dom";
import App from "./App.js";
import { chooseAllCharactersIn } from "./util/ChoosingCharacters.js";
import getFirebaseFunctions from "./util/firebase.js";

jest.mock("./components/HighscoreScreen.js", () => () => {
  return <div data-testid="HighscoreScreen"></div>;
});

jest.mock("./components/Stopwatch.js", () => () => {
  return <div></div>;
});

beforeAll(async () => {
  const { addFakeCharacterCoordsToDatabase, clearDatabase } = getFirebaseFunctions()
  await clearDatabase()
  await addFakeCharacterCoordsToDatabase()
})

describe.each(['maze', 'beach', 'snow'])('for %s,', (mapName) => {
  it("initially Gamescreen is shown and HighscoreScreen is hidden", () => {
    render(
      <MemoryRouter initialEntries={[`/${mapName}`]}>
        <App />
      </MemoryRouter>
    );

    expect(screen.getByTestId("Gamescreen")).toBeDefined();
    expect(screen.queryByTestId("HighscoreScreen")).toBeNull();
  });

  test("Shows highscore screen and hides gamescreen when all characters are chosen", async () => {
    render(
      <MemoryRouter initialEntries={[`/${mapName}`]}>
        <App />
      </MemoryRouter>
    );

    await chooseAllCharactersIn(mapName);

    expect(screen.queryByTestId("Gamescreen")).toBeNull();
    expect(screen.getByTestId("HighscoreScreen")).toBeDefined();
  });
})

