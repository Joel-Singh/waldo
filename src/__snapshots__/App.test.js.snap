// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selection screen is passed map previews 1`] = `
Array [
  Array [
    Object {
      "mapPreviews": Array [
        <MapPreview
          mapImage="maze.jpg"
          mapName="Maze"
          mapPath="maze"
          previewCharacterInformation={
            Array [
              Object {
                "image": "waldo.jpg",
                "name": "Waldo",
              },
              Object {
                "image": "bird_person.jpg",
                "name": "Bird Person",
              },
              Object {
                "image": "yellowed_hair_person.jpg",
                "name": "Yellow Hair Person",
              },
            ]
          }
        />,
        <MapPreview
          mapImage="beach.jpg"
          mapName="Beach"
          mapPath="beach"
          previewCharacterInformation={
            Array [
              Object {
                "image": "waldo.jpg",
                "name": "Waldo",
              },
              Object {
                "image": "shovel_boy.jpg",
                "name": "Shovel Boy",
              },
              Object {
                "image": "dog.jpg",
                "name": "Dog",
              },
            ]
          }
        />,
        <MapPreview
          mapImage="snow.jpg"
          mapName="Snow"
          mapPath="snow"
          previewCharacterInformation={
            Array [
              Object {
                "image": "waldo.jpg",
                "name": "Waldo",
              },
              Object {
                "image": "ouch.jpg",
                "name": "Ouch!",
              },
              Object {
                "image": "monster.jpg",
                "name": "Monster",
              },
            ]
          }
        />,
      ],
    },
    Object {},
  ],
]
`;
